dash/development/_r_components_generation.py

```diff
@@ -0,0 +1,511 @@
+
+    function_string = ''.join([function_frame_open,
+                               function_frame_body,
+                               frame_close_template])
+
+    return function_string
+
+
+def write_help_file_r(name, props, description, prefix):
+    """
+    Write R documentation file (.Rd) given component name and properties
+
+    Parameters
+    ----------
+    name
+    props
+    description
+    prefix
+
+    Returns
+    -------
+
+
+    """
+    file_name = '{}{}.Rd'.format(prefix, name)
+    prop_keys = props.keys()
+
+    default_argtext = ''
+    item_text = ''
+
+    # Ensure props are ordered with children first
+    props = reorder_props(props=props)
+
+    default_argtext += ", ".join(
+        '{}={}'.format(p, json_to_r_type(props[p]))
+        if 'defaultValue' in props[p] else
+        '{}=NULL'.format(p)
+        for p in prop_keys
+        if not p.endswith("-*") and
+        p not in r_keywords and
+        p not in ['setProps', 'dashEvents', 'fireEvent']
+    )
+
+    item_text += "\n\n".join(
+        '\\item{{{}}}{{{}}}'.format(p, props[p]['description'])
+        for p in prop_keys
+        if not p.endswith("-*") and
+        p not in r_keywords and
+        p not in ['setProps', 'dashEvents', 'fireEvent']
+    )
+
+    file_path = os.path.join('man', file_name)
+    with open(file_path, 'w') as f:
+        f.write(help_string.format(
+            prefix=prefix,
+            name=name,
+            default_argtext=default_argtext,
+            item_text=item_text,
+            description=description.replace('\n', ' ')
+        ))
+
+
+def write_class_file_r(name,
+                       props,
+                       description,
+                       project_shortname,
+                       prefix=None):
+    import_string =\
+        "# AUTO GENERATED FILE - DO NOT EDIT\n\n"
+    class_string = generate_class_string_r(
+        name,
+        props,
+        project_shortname,
+        prefix
+    )
+    file_name = "{}{}.R".format(prefix, name)
+
+    file_path = os.path.join('R', file_name)
+    with open(file_path, 'w') as f:
+        f.write(import_string)
+        f.write(class_string)
+
+    # generate the R help pages for each of the Dash components that we
+    # are transpiling -- this is done to avoid using Roxygen2 syntax,
+    # we may eventually be able to generate similar documentation using
+    # doxygen and an R plugin, but for now we'll just do it on our own
+    # from within Python
+    # noqa E344
+    write_help_file_r(
+        name,
+        props,
+        description,
+        prefix
+    )
+
+    print('Generated {}'.format(file_name))
+
+
+# pylint: disable=inconsistent-return-statements
+def generate_export_string_r(name, prefix):
+    if not name.endswith('-*') and \
+            str(name) not in r_keywords and \
+            str(name) not in ['setProps', 'children', 'dashEvents']:
+        return 'export({}{})\n'.format(prefix, name)
+
```

Comment:
That's a good question, unfortunately my understanding of the wildcards is limited. From conversations with @T4rk1n, it's unclear whether they are used very often, so I have no easy way to test the repercussions.

Can you point me at examples other than [this one](https://github.com/plotly/dash/blob/75a285f16da29780a3658d9d92a9847c15e7549d/tests/test_integration.py#L178) where wildcards are used?

Reply:
The `generate_export_string` function was used to write the `NAMESPACE` file for the R package. It returned the name of the R functions generated by the JSON >> R 'transpiler' for each of the Dash components.

We've since relocated this code within `generate_exports`.